### ISSUE TITLE
Go API: too many dependencies

### ISSUE THREAD

**Issue by dolmen** [hashicorp/vault#14017]
**Describe the bug**

Very long list of dependencies when importing package `github.com/hashicorp/vault/api` which is supposed to be just a wrapper around a REST API.

Here is a list of package that I find completely irrelevant to the purpose of the imported package:
* [`github.com/fatih/color`](https://pkg.go.dev/github.com/fatih/color)
* [`github.com/mattn/go-colorable`](https://pkg.go.dev/github.com/mattn/go-colorable)
* [`github.com/mattn/go-isatty`](https://pkg.go.dev/github.com/mattn/go-isatty)
* ....

**To Reproduce**
```go
package main

import _ "github.com/hashicorp/vault/api"

func main() {
}
```
```console
$ go mod init tmp
$ go mod tidy
$ go build -o tmp
tmp: go1.17.6
	path	tmp
	mod	tmp	(devel)	
	dep	github.com/armon/go-metrics	v0.3.9	h1:O2sNqxBdvq8Eq5xmzljcYzAORli6RWCvEym4cJf9m18=
	dep	github.com/armon/go-radix	v1.0.0	h1:F4z6KzEeeQIMeLFa97iZU6vupzoecKdU5TX24SNppXI=
	dep	github.com/cenkalti/backoff/v3	v3.0.0	h1:ske+9nBpD9qZsTBoF41nW5L+AIuFBKMeze18XQ3eG1c=
	dep	github.com/fatih/color	v1.7.0	h1:DkWD4oS2D8LGGgTQ6IvwJJXSL5Vp2ffcQg58nFV38Ys=
	dep	github.com/golang/protobuf	v1.5.2	h1:ROPKBNFfQgOUMifHyP+KYbvpjbdoFNs+aK7DXlji0Tw=
	dep	github.com/golang/snappy	v0.0.4	h1:yAGX7huGHXlcLOEtBnF4w7FQwA26wojNCwOYAEhLjQM=
	dep	github.com/hashicorp/errwrap	v1.1.0	h1:OxrOeh75EUXMY8TBjag2fzXGZ40LB6IKw45YeGUDY2I=
	dep	github.com/hashicorp/go-cleanhttp	v0.5.2	h1:035FKYIWjmULyFRBKPs8TBQoi0x6d9G4xc9neXJWAZQ=
	dep	github.com/hashicorp/go-hclog	v0.16.2	h1:K4ev2ib4LdQETX5cSZBG0DVLk1jwGqSPXBjdah3veNs=
	dep	github.com/hashicorp/go-immutable-radix	v1.3.1	h1:DKHmCUm2hRBK510BaiZlwvpD40f8bJFeZnpfm2KLowc=
	dep	github.com/hashicorp/go-multierror	v1.1.1	h1:H5DkEtf6CXdFp0N0Em5UCwQpXMWke8IA0+lD48awMYo=
	dep	github.com/hashicorp/go-plugin	v1.4.3	h1:DXmvivbWD5qdiBts9TpBC7BYL1Aia5sxbRgQB+v6UZM=
	dep	github.com/hashicorp/go-retryablehttp	v0.6.6	h1:HJunrbHTDDbBb/ay4kxa1n+dLmttUlnP3V9oNE4hmsM=
	dep	github.com/hashicorp/go-rootcerts	v1.0.2	h1:jzhAVGtqPKbwpyCPELlgNWhE1znq+qwJtW5Oi2viEzc=
	dep	github.com/hashicorp/go-secure-stdlib/mlock	v0.1.1	h1:cCRo8gK7oq6A2L6LICkUZ+/a5rLiRXFMf1Qd4xSwxTc=
	dep	github.com/hashicorp/go-secure-stdlib/parseutil	v0.1.1	h1:78ki3QBevHwYrVxnyVeaEz+7WtifHhauYF23es/0KlI=
	dep	github.com/hashicorp/go-secure-stdlib/strutil	v0.1.1	h1:nd0HIW15E6FG1MsnArYaHfuw9C2zgzM8LxkG5Ty/788=
	dep	github.com/hashicorp/go-sockaddr	v1.0.2	h1:ztczhD1jLxIRjVejw8gFomI1BQZOe2WoVOu0SyteCQc=
	dep	github.com/hashicorp/go-uuid	v1.0.2	h1:cfejS+Tpcp13yd5nYHWDI6qVCny6wyX2Mt5SGur2IGE=
	dep	github.com/hashicorp/go-version	v1.2.0	h1:3vNe/fWF5CBgRIguda1meWhsZHy3m8gCJ5wx+dIzX/E=
	dep	github.com/hashicorp/golang-lru	v0.5.4	h1:YDjusn29QI/Das2iO9M0BHnIbxPeyuCHsjMW+lJfyTc=
	dep	github.com/hashicorp/hcl	v1.0.0	h1:0Anlzjpi4vEasTeNFn2mLJgTSwt0+6sfsiTG8qcWGx4=
	dep	github.com/hashicorp/vault/api	v1.3.1	h1:pkDkcgTh47PRjY1NEFeofqR4W/HkNUi9qIakESO2aRM=
	dep	github.com/hashicorp/vault/sdk	v0.3.0	h1:kR3dpxNkhh/wr6ycaJYqp6AFT/i2xaftbfnwZduTKEY=
	dep	github.com/hashicorp/yamux	v0.0.0-20180604194846-3520598351bb	h1:b5rjCoWHc7eqmAS4/qyk21ZsHyb6Mxv/jykxvNTkU4M=
	dep	github.com/mattn/go-colorable	v0.1.6	h1:6Su7aK7lXmJ/U79bYtBjLNaha4Fs1Rg9plHpcH+vvnE=
	dep	github.com/mattn/go-isatty	v0.0.12	h1:wuysRhFDzyxgEmMf5xjvJ2M9dZoWAXNNr5LSBS7uHXY=
	dep	github.com/mitchellh/copystructure	v1.0.0	h1:Laisrj+bAB6b/yJwB5Bt3ITZhGJdqmxquMKeZ+mmkFQ=
	dep	github.com/mitchellh/go-homedir	v1.1.0	h1:lukF9ziXFxDFPkA1vsr5zpc1XuPDn/wFntq5mG+4E0Y=
	dep	github.com/mitchellh/go-testing-interface	v1.0.0	h1:fzU/JVNcaqHQEcVFAKeR41fkiLdIPrefOvVG1VZ96U0=
	dep	github.com/mitchellh/mapstructure	v1.4.2	h1:6h7AQ0yhTcIsmFmnAwQls75jp2Gzs4iB8W7pjMO+rqo=
	dep	github.com/mitchellh/reflectwalk	v1.0.0	h1:9D+8oIskB4VJBN5SFlmc27fSlIBZaov1Wpk/IfikLNY=
	dep	github.com/oklog/run	v1.0.0	h1:Ru7dDtJNOyC66gQ5dQmaCa0qIsAUFY3sFpK1Xk8igrw=
	dep	github.com/pierrec/lz4	v2.5.2+incompatible	h1:WCjObylUIOlKy/+7Abdn34TLIkXiA4UWUMhxq9m9ZXI=
	dep	github.com/ryanuber/go-glob	v1.0.0	h1:iQh3xXAumdQ+4Ufa5b25cRpC5TYKlno6hsv6Cb3pkBk=
	dep	go.uber.org/atomic	v1.9.0	h1:ECmE8Bn/WFTYwEW/bpKD3M8VtR/zQVbavAoalC1PYyE=
	dep	golang.org/x/crypto	v0.0.0-20210711020723-a769d52b0f97	h1:/UOmuWzQfxxo9UtlXMwuQU8CMgg1eZXqTRwkSQJWKOI=
	dep	golang.org/x/net	v0.0.0-20210226172049-e18ecbb05110	h1:qWPm9rbaAMKs8Bq/9LRpbMqxWRVUAQwMI9fVrssnTfw=
	dep	golang.org/x/sys	v0.0.0-20210630005230-0f9fa26af87c	h1:F1jZWGFhYfh0Ci55sIpILtKKK8p3i2/krTr0H1rg74I=
	dep	golang.org/x/text	v0.3.3	h1:cokOdA+Jmi5PJGXLlLllQSgYigAEfHXJAERHVMaCc2k=
	dep	golang.org/x/time	v0.0.0-20200416051211-89c76fbcd5d1	h1:NusfzzA6yGQ+ua51ck7E3omNUX/JuqbFSaRGqU8CcLI=
	dep	google.golang.org/genproto	v0.0.0-20200526211855-cb27e3aa2013	h1:+kGHl1aib/qcwaRi1CbqBZ1rk19r85MNUf8HaBghugY=
	dep	google.golang.org/grpc	v1.41.0	h1:f+PlOh7QV4iIJkPrx5NQ7qaNGFQ3OTse67yaDHfju4E=
	dep	google.golang.org/protobuf	v1.26.0	h1:bxAC2xTBsZGibn2RTntX0oH50xLsqy1OxA9tTL3p/lk=
	dep	gopkg.in/square/go-jose.v2	v2.5.1	h1:7odma5RETjNHWJnR32wx8t+Io4djHE1PqxCFx3iiZ2w=
$ go mod why github.com/mattn/go-isatty
# github.com/mattn/go-isatty
tmp
github.com/hashicorp/vault/api
github.com/hashicorp/vault/sdk/logical
github.com/hashicorp/go-hclog
github.com/mattn/go-isatty
```

**Expected behavior**
A short list of dependencies. No packages that are obivously related to terminal features.

**Environment:**
`GOOS=darwin GOARCH=amd64`

**Comment by dolmen** [comment]
[On the Go playground](https://go.dev/play/p/hlzwqcIhT90), compilation fails (`timeout running go build`).

**Comment by dolmen** [comment]
The only reference to package `github.com/hashicorp/vault/sdk/logical` comes from the following function in client.go:

`func `[`ParseReplicationState`](https://pkg.go.dev/github.com/hashicorp/vault/api#ParseReplicationState)`(raw string, hmacKey []byte) (*`[`logical`](https://pkg.go.dev/github.com/hashicorp/vault/sdk/logical)`.`[`WALState`](https://pkg.go.dev/github.com/hashicorp/vault/sdk/logical#WALState)`, error)`

**Comment by hsimon-hashicorp** [comment]
Hi @dolmen! The API is meant to accommodate a wide range of use cases, which means that the dependencies list can get quite long. This is by design. For example, the API depends on sdk for some things (types, constants, etc), and the sdk depends on hclog, among others. I'm going to go ahead and close this issue now, but we appreciate your input. :)

**Comment by dolmen** [comment]
Ok. Thanks. This API is now unusable. I will switch to just core `net/http`.

**Comment by dolmen** [comment]
For reference: [Our Software Dependency Problem](https://research.swtch.com/deps), by Russ Cox.

**Comment by dolmen** [comment]
Here is [a PR for `hclog`](https://github.com/hashicorp/go-hclog/pull/106 ) that will help to remove one dependency.

**Comment by dolmen** [comment]
Commit 2fe214f5719a17b67f2630f82f8e51797db1a633 is adding one more link to sdk/logical.

**Comment by dolmen** [comment]
@hsimon-hashicorp A wide range of uses cases is not incompatible with a cleaner implementation with less dependencies. See #14150.

So please reopen this issue.

**Comment by VinnyHC** [comment]
We understand the frustration around our current client libraries and are actively working to create a more streamlined experience with the introduction of new libraries for Vault. As @hsimon-hashicorp mentioned, the current API package is designed to handle various use-cases, including some internal to Vault. While this may not be ideal for all developers, it has proven to be a reasonable compromise. Your feedback is always appreciated while we work to continuously improve Vault and its developer experience.

**Comment by dolmen** [comment]
This issue is being addressed in #18962. Thanks @ncabatoff!

**Comment by dolmen** [comment]
I consider that the issue is fixed in `github.com/hashicorp/vault/api@v1.9.0`.
